### ISSUE TITLE
Value retrieval slowdown for payloads over 5 MB with SSL transport

### ISSUE THREAD

**Issue by Joas1988** [redis/redis-rb#935]
To replicate connect to redis using SSL and test set and get with payload of more than 5 MB.

If you attach an strace to a process that is trying to do a large redis read, you can see that it is in a loop of

- allocate a buffer of the length of the [bulk string](https://github.com/redis/redis-rb/blob/d75e074c9997f8d099b6a0f65f0cf517274943f2/lib/redis/connection/ruby.rb#L38). The allocation in this case is an mmap - as malloc normally uses mmap when trying to alloc giant memory sizes.

- syscall to read from the socket - usually around 1400 bytes or so

- repeat until we are done

It boils down to that Ruby OpenSSL is reading 1400 bytes and without using buffer it [allocates buffer internally](https://github.com/ruby/openssl/blob/41587f69e17b9f0983c1f2a37b8661599119fc0e/ext/openssl/ossl_ssl.c#L1834) for length provided as argument to [def read_nonblock](https://github.com/ruby/openssl/blob/master/lib/openssl/buffering.rb#L195.) In my company we solved that issue by this patch.

```ruby 
Redis::Connection::SocketMixin.module_eval do
    def read(nbytes)
      result = @buffer.slice!(0, nbytes)
      # always use this pre-allocated buffer when calling into openssl
      buffer = String.new(capacity: nbytes, encoding: Encoding::ASCII_8BIT)
      result << _read_from_socket(nbytes - result.bytesize, buffer) while result.bytesize < nbytes
      result
    end

    def _read_from_socket(nbytes, buffer = nil)
      # addition here to use the pre allocated buffer
      read_nonblock(nbytes, buffer)
    rescue IO::WaitReadable
      if IO.select([self], nil, nil, @timeout)
        retry
      else
        raise Redis::TimeoutError
      end
    rescue IO::WaitWritable
      if IO.select(nil, [self], nil, @timeout)
        retry
      else
        raise Redis::TimeoutError
      end
    end
  end
```

**Comment by rbroemeling** [comment]
@Joas1988 Thanks for discovering this and posting this patch -- we encountered exactly this problem in our production environment with larger result sizes (i.e., results of a few MB or higher). It was very noticeable, and your patch mitigated the impact greatly.

Have you heard anything back from the maintainers of `redis-rb` on this topic? This seems like a serious problem in the library, so I'm a little curious at the lack of activity or interest on this change? Any ideas?

**Comment by rbroemeling** [comment]
@byroot / @godric I see that the two of you are active in the repository -- is there anything that we/I can do to help this change get merged into the redis-rb gem so that this performance problem is fixed?

**Comment by byroot** [comment]
Well, first step could be to convert this monkey patch into a proper pull request.

**Comment by byroot** [comment]
Fixed in #1023 / fe18b814628a95ac0addc2064caf15b54ca7c004. Thanks for the issue.
